src/Controller/PricingAndPlansController.php

```diff
@@ -107,63 +130,53 @@ public function catalogPage(UserInterface $user) {
-    $packages = Package::getAvailableApiPackagesByDeveloper($user->getEmail());
-    // Get the view mode from package config.
-    $view_mode = $this->config(PackageConfigForm::CONFIG_NAME)->get('catalog_view_mode');
-    $build = ['package_list' => $this->entityTypeManager()->getViewBuilder('package')->viewMultiple($packages, $view_mode ?? 'default')];
-    $build['package_list']['#pre_render'][] = [$this, 'preRender'];
 
-    return $build;
-  }
+    $rate_plans = [];
 
-  /**
-   * Get a rate plan controller.
-   *
-   * @param string $package_id
-   *   The package ID.
-   *
-   * @return \Apigee\Edge\Api\Monetization\Controller\RatePlanControllerInterface
-   *   The rate plan controller.
-   */
-  protected function packageRatePlanController($package_id): RatePlanControllerInterface {
-    return $this->controller_factory->ratePlanController($package_id);
+    // Load rate plans for each package.
+    foreach (Package::getAvailableApiPackagesByDeveloper($user->getEmail()) as $package) {
+      /** @var \Drupal\apigee_m10n\Entity\PackageInterface $package */
+      foreach ($package->get('ratePlans') as $rate_plan) {
+        $rate_plans["{$package->id()}:{$rate_plan->target_id}"] = $rate_plan->entity;
+      };
+    }
+
+    return $this->buildPage($rate_plans);
   }
 
   /**
-   * Call back to pre-process the entity list.
+   * Builds the page for a rate plan listing.
    *
-   * @param array $build
-   *   A render array as processed by
-   *   `\Drupal\Core\Entity\EntityViewBuilder::buildMultiple()`.
+   * @param \Drupal\apigee_m10n\Entity\RatePlanInterface[] $plans
+   *   A list of rate plans.
    *
    * @return array
-   *   The processed render array.
+   *   A render array for plans.
    */
-  public function preRender(array $build) {
-    // Add a wrapper around the list.
-    $build['#prefix'] = '<ul class="apigee-package-list">';
-    $build['#suffix'] = '</ul>';
-    // Wrap each package as a list item.
-    foreach (Element::children($build) as $index) {
-      // Header HTML.
-      $header_html = '
-        <div class="apigee-sdk-package-basic">
-          <div class="apigee-package-label">@package_label</div>
-          <div class="apigee-sdk-product-list-basic">(<span class="apigee-sdk-product"> @product_labels </span>)</div>
-        </div>
-        <div class="apigee-package-details">';
-      // Build a list of product labels for the basic info header.
-      $product_labels = [];
-      foreach ($build[$index]['#package']->get('apiProducts') as $item) {
-        $product_labels[] = $item->entity->label();
-      }
-      $build[$index]['#prefix'] = new FormattableMarkup('<li class="apigee-sdk-package">' . $header_html, [
-        '@package_label' => $build[$index]['#package']->label(),
-        '@product_labels' => implode(', ', $product_labels),
-      ]);
-      $build[$index]['#suffix'] = '</div></li>';
+  protected function buildPage($plans) {
+    $build = [
+      '#type' => 'container',
```

Comment:
@Jaesin Can you please add this line? We need to be able to target that container.

`'#theme' => ['container__rate_plan_list'],`

Thanks!

Reply:
@Jaesin Actually, apparently `#theme_wrappers` is better in this case, because of processing that happens.  Using `#theme` apparently results in [another rendering step](https://git.drupalcode.org/project/drupal/blob/8.8.x/core/lib/Drupal/Core/Render/RendererInterface.php#L204-219), and`{{ children }}` (which is what the default template implementation uses) no longer works because it's already rendered as #markup and lives under `{{ element }}`.  Using `#theme_wrappers` doesn't change that pipeline and gives the ability to customize as we need:

`'#theme_wrappers' => ['container__rate_plan_list'],`